contextual.go

```diff
@@ -0,0 +1,186 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package klog
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+)
+
+// This file provides the implementation of
+// https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging
+//
+// SetLogger and ClearLogger were originally added to klog.go and got moved
+// here. Contextual logging adds a way to retrieve a Logger for direct logging
+// without the logging calls in klog.go.
+//
+// The global variables are expected to be modified only during sequential
+// parts of a program (init, serial tests) and therefore are not protected by
+// mutex locking.
+
+var (
+	// contextualLoggingEnabled controls whether contextual logging is
+	// active. Disabling it may have some small performance benefit.
+	contextualLoggingEnabled = true
+
+	// globalLogger is the global Logger chosen by users of klog, nil if
+	// none is available.
+	globalLogger *Logger
+
+	// klogLogger is used as fallback for logging through the normal klog code
+	// when no Logger is set.
+	klogLogger logr.Logger = logr.New(&klogger{})
+)
+
+// SetLogger sets a Logger implementation that will be used in two different
+// ways:
+// - as backing implementation of the traditional klog log calls
+// - for direct log calls afer retrieving it with FromContext, TODO, or
```

Comment:
I need to call out one thing here: this requirement that direct log calls are also okay is new. Some loggers that were installed with SetLogger for use as backend might not work as expected because they don't check verbosity.

For example, the original JSON logger from `component-base/logs` didn't check verbosity and instead relied entirely on klog to do that for it. If called directly, all log entries would have gotten emitted.

Given that SetLogger is still relatively new, I feel that this change of semantics is acceptable. But if we want to be absolutely sure about not breaking anyone, we can also do this instead:
- keep SetLogger as it is (i.e. loggers will only be called as backend)
- add a new SetContextualLogger (same semantic as described here)
- both functions set globalLogger, but also a boolean
- Background returns klogr (i.e. logging goes through klog) when SetLogger was used, otherwise the logger

Reply:
@serathius pointed out that `SetLogger` actually predates structured logging and was used already earlier to redirect Kubernetes output into other loggers.

[A code search](https://grep.app/search?q=klog.SetLogger) indeed finds users who rely on verbosity filtering in klog:
- https://github.com/projectcontour/contour/blob/b9413e8851be0ff8dfb230ef872a015b168e990b/internal/k8s/log.go#L78-L84
- https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/07b6ace393f32fd57a0c1f51e62155e4c2fc499a/cmd/main.go#L56 (the old JSON logger without verbosity)

Let's play it safe and use `SetContextualLogger`.